File: day15/day15.cpp
#include <algorithm>
#include <cassert>
#include <cstdint>
#include <iostream>
#include <unordered_map>
#include <deque>

template <typename T>
void output(T v) {
	std::cout << v << std::endl;
}


uint32_t get_up_to(std::initializer_list<uint32_t> const& numbers, uint64_t lim)
{
	std::deque<uint32_t> times(4096);
	//std::unordered_map<uint32_t, uint64_t> times(lim); // / 16);
	uint64_t time = 0;
	uint32_t n;
	for (; time < numbers.size(); ++time) {
		n = numbers.begin()[time];
		times[n] = time + 1;
	}
	for (; time < (lim); ++time) { // lim -1?	
		if (0 == (time & 0x0FFFFF)) {
			output(time);
		}
		auto const last = n;
		if (n > times.size()) {
			times.resize(n + 1);
		}
		auto found = times[n];
		if (found == 0) {
			n = 0;
		}
		else {
			n = static_cast<uint32_t>(time - found);
		}
		times[last] = time;
	}
	return n;
}

void main() {
	auto t = get_up_to({0,3,6}, 2020);
	output(t);
	//assert(t == 436);
	auto p1 = get_up_to({0,14,6,20,1,4}, 2020);

	//assert(p1 == 257);
	output(p1);

	auto p2 = get_up_to({0,14,6,20,1,4}, 30000000);
	output(p2);
}

File: day15/day15.py
import os
import copy

sdir = os.path.dirname(os.path.realpath(__file__))

data = [0,14,6,20,1,4]

def get_up_to(numbers, lim):
	times={}
	for i,n in enumerate(numbers):
		times.update({n: i})

	time = len(numbers) - 1
	n = numbers[time]
	for time in range(len(numbers)-1, lim-1):
		last = n
		if n in times:
			n = time - times[n]
		else:
			n = 0
		times.update({last:time})
		time += 1

		if (0 == time % 100000):
			print(time)
		#print(time, ", ", n)
		# 0 3 3 1 0 4 0

	return n


t = get_up_to([0,3,6], 2020)
assert(t == 436)
p1 = get_up_to(data, 2020)
assert(p1 == 257)
print(p1)
p2 = get_up_to(data, 30000000)
print(p2)
